### PR TITLE
#4665: support AA Community Edition on partner onboarding screen

### PR DESCRIPTION
**Pull request by twschiller**
## What does this PR do?

- Closes #4665 

## Reviewer Notes

- Corresponds to app PR: https://github.com/pixiebrix/pixiebrix-app/pull/2627. The app PR is pinned to this commit to have the partner flow through
- Start by reviewing installer.ts

## Demo

- https://www.loom.com/share/06261edd5f984fc8aa8c26add8a3d21d

## Remaining Work

- [x] Show CE AARI onboarding on first run
- [x] Support CE after extension is linked to account

## Future Areas of Improvement

- The transition between Admin Console to Extension Console is awkward (see Loom). Consider showing a loading/initialization screen in the Admin Console instead of the default extension page
- Auto-fill the Community Edition Control Room URL. Would need to save it in installer.ts:openInstallPage

## Checklist

- [x] Add tests
- [x] Run Storybook and manually confirm that all stories are working
- [x] Designate a primary reviewer: @mnholtz

### CODE REVIEW

File: src/auth/token.ts [hunk 1]
```diff
@@ -182,16 +182,16 @@ export async function clearCachedAuthSecrets(): Promise<void> {
  * @see linkExtension
  */
 export async function updateUserData(update: UserDataUpdate): Promise<void> {
-  const updated = await readAuthData();
+  const result = await readAuthData();
```
Comment (twschiller) [inline src/auth/token.ts:185]: Switch variable name to one that is easier to distinguish from the parameter name

File: src/auth/authTypes.ts [hunk 2]
```diff
@@ -72,11 +72,23 @@ export type UserData = Partial<{
...
    */
   enforceUpdateMillis: number | null;
+  /**
+   * The partner, controlling theme, documentation links, etc.
+   *
+   * Introduced to support partner Community Edition users because they are not part of an organization.
+   *
+   * @since 1.7.14
+   */
+  readonly partner?: Me["partner"];
```
Comment (twschiller) [inline src/auth/authTypes.ts:82]: Pass partner from the app when linking the extension

File: src/auth/useRequiredPartnerAuth.ts [hunk 3]
```diff
@@ -86,22 +86,22 @@ export type RequiredPartnerState = {
 
 function decidePartnerServiceIds({
   authServiceIdOverride,
-  authMethod,
+  authMethodOverride,
```
Comment (twschiller) [inline src/auth/useRequiredPartnerAuth.ts:89]: Clarify parameter name

File: src/background/installer.ts [hunk 4]
```diff
@@ -34,51 +39,181 @@ const SERVICE_URL = process.env.SERVICE_URL;
...
+  expectContext("background");
+
+  // Look for an Admin Console tab that's showing an onboarding page
+  // /setup: normal onboarding screen
+  // /start: partner onboarding
+  const [appBaseUrl, [appOnboardingTab]] = await Promise.all([
     getBaseURL(),
     browser.tabs.query({
+      // Can't use SERVICE_URL directly because it contains a port number during development, resulting in an
+      // invalid URL match pattern
```
Comment (mnholtz) [inline src/background/installer.ts:55]: Do you have an example? I've never had this problem before.

File: src/background/installer.test.ts [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+        id: 1,
+        url: `${APP_BASE_URL}/setup`,
+      },
+    ]);
+    await openInstallPage();
+    expect(updateTabMock).toHaveBeenCalledWith(1, {
+      url: APP_BASE_URL,
+      active: true,
+    });
+    expect(createTabMock.mock.calls.length).toBe(0);
```
Comment (mnholtz) [inline src/background/installer.test.ts:71]: 👍 ❤️

File: src/background/installer.ts [hunk 6]
```diff
@@ -34,51 +39,181 @@ const SERVICE_URL = process.env.SERVICE_URL;
...
 }
 
-async function install({ reason }: Runtime.OnInstalledDetailsType) {
+/**
+ * For partner installs, if a partner integration is not configured, automatically open the Extension Console
+ * to continue the partner authentication onboarding flow.
+ *
+ * @see useRequiredPartnerAuth
+ */
+export async function checkPartnerAuth(): Promise<void> {
```
Comment (mnholtz) [inline src/background/installer.ts:144]: Nit: can we call this something else to emphasize the side effect? Like maybe `requirePartnerIntegration` or `checkPartnerAuthEffect`?

File: src/background/messenger/external/_implementation.ts [hunk 7]
```diff
@@ -23,16 +23,22 @@ import { linkExtension } from "@/auth/token";
 import { TokenAuthData } from "@/auth/authTypes";
 import { reportEvent } from "@/telemetry/events";
 
+const HACK_EXTENSION_LINK_RELOAD_DELAY_MS = 100;
```
Comment (mnholtz) [inline src/background/messenger/external/_implementation.ts:26]: ❤️

**Review by mnholtz — APPROVED** [review]

**Comment by mnholtz** [comment]
Note that I also tested these onboarding flows in conjunction with the app-side change: https://github.com/pixiebrix/pixiebrix-app/pull/2627#issuecomment-1322631031

File: src/options/pages/onboarding/partner/PartnerSetupCard.tsx [hunk 8]
```diff
@@ -81,28 +96,17 @@ function usePartnerLoginMode(): "token" | "oauth2" {
...
+      return hasCachedNativeLogin || isCommunityControlRoom(hostname)
+        ? "token"
+        : "oauth2";
     }
   }
 }
 
 const CONTROL_ROOM_URL_MANAGED_KEY = "controlRoomUrl" as ManualStorageKey;
 
-function hostnameToUrl(hostname: string): string {
```
Comment (johnnymetz) [inline src/options/pages/onboarding/partner/PartnerSetupCard.tsx:93]: Reviewer note: moved to `aaUtils.ts`

**Review by johnnymetz — APPROVED** [review]

File: src/background/installer.ts (hunk 4 above)
Comment (twschiller) [reply src/background/installer.ts:55]: > Do you have an example? I've never had this problem before.

I was getting it with http://127.0.0.1:8000

**Pull request merged**